### ISSUE TITLE
main.py

### ISSUE THREAD

**Issue by LilyDreamZhao** [LilyDreamZhao/DeepLearning#2]
# coding: utf-8
import tensorflow as tf
import gc
import psutil
import os
from DataClass import DataSet
import numpy as np
from skimage import img_as_float  
import cv2
import argparse
import sys

parser = argparse.ArgumentParser()

parser.add_argument('-n', action='store', default=45,dest='n_classes',type=int,
        help='input numbers of classes (2-45,default=45)')
		
parser.add_argument('-t', action='store', default=500,dest='num_step',type=int,
        help='input numbers of classes (default=45)')
		
parser.add_argument('-b', action='store', default=1,dest='batch_size',type=int,
        help='input numbers of classes (default=1)')
parser.add_argument('-p', action='store', default=1,dest='perprint',type=int,
        help='input p fo printint per p steps(default=1)')
results = parser.parse_args()
n_classes = results.n_classes
num_step = results.num_step
batch_size = results.batch_size
perprint = results.perprint

path = sys.path[0]+"\\NWPU-RESISC45"

def getData():
    files= os.listdir(path) #Get all file names under the folder
    X_allclasses=[]
    i=1
    for file in files: 
        if i>n_classes:
           break
        X_oneclasses=[]
        pictures=os.listdir(path+"/"+file)
        for picture in pictures:
            img=cv2.imread(path+"/"+file+"/"+picture)
            X_oneclasses.append(img)
        X_allclasses.append(X_oneclasses)
        i=i+1
    return np.array(X_allclasses)
    print("get data finished...")

def one_hot(ys_):
    ys=np.zeros((ys_.shape[0],n_classes))
    for i in np.arange(ys_.shape[0]):
        j = int(ys_[i])
        ys[i,j-1] = 1
    return ys
def removehead(data):
    return data[:,1:]

#初始化权值
def weight_variable(shape):
    initial = tf.truncated_normal(shape,stddev=0.1)#生成一个截断的正态分布
    return tf.Variable(initial)

#初始化偏置
def bias_variable(shape):
    initial = tf.constant(0.1,shape=shape)
    return tf.Variable(initial)

#卷积层
def conv2d(x,W,strides_value):
    #x input tensor of shape `[batch, in_height, in_width, in_channels]`
    #W filter / kernel tensor of shape [filter_height, filter_width, in_channels, out_channels]
    #`strides[0] = strides[3] = 1`. strides[1]代表x方向的步长，strides[2]代表y方向的步长
    #padding: A `string` from: `"SAME", "VALID"`
    return tf.nn.conv2d(x,W,strides=[1,strides_value,strides_value,1],padding='SAME')

#池化层
def max_pool(x,ksize_value,strides_value):
    #ksize [1,x,y,1]
    return tf.nn.max_pool(x,ksize=[1,ksize_value,ksize_value,1],strides=[1,strides_value,strides_value,1],padding='SAME')
	
	
def runModel():
	data=getData()
	data=img_as_float(data)
	size=256
	n_channel=3

	data=np.reshape(data,[n_classes*700,256,256,3])
	data=data[:,0:size,0:size,0:n_channel]
	data=np.reshape(data,[n_classes*700,size*size*n_channel])
	dataX=data.astype(np.float64)

	print(dataX.dtype)

	del data
	data=None
	gc.collect()

	data=dataX
	Y=[]
	for i in range(n_classes):
		y=[i]*700
		Y.extend(y)
	data = np.insert(data, 0, values=Y, axis=1)

	index_num_example=np.arange(n_classes*700)
	np.random.shuffle(index_num_example)
	data = data[index_num_example]

	data_train=data[0:data.shape[0]//3*2]
	data_test=data[data.shape[0]//3*2:]

	X_train = removehead(data_train)
	Y_train = one_hot(data_train[:,0])
	X_test =  removehead(data_test)
	Y_test = one_hot(data_test[:,0])
	train_p = DataSet(data_train)
	train_p2 = DataSet(data_train)
	test_p = DataSet(data_test)

	x = tf.placeholder(tf.float32,[None,size*size*n_channel])
	y = tf.placeholder(tf.float32,[None,n_classes])
	x_image = tf.reshape(x,[-1,size,size,n_channel])

	#1
	W_conv_1 = weight_variable([11,11,n_channel,96])#5*5的采样窗口，32个卷积核从1个平面抽取特征
	b_conv_1 = bias_variable([96])#每一个卷积核一个偏置值
	conv2d_1 = conv2d(x_image,W_conv_1,4) + b_conv_1
	relu_1 = tf.nn.relu(conv2d_1)

	#2
	h_pool_2 = max_pool(relu_1,3,2)#进行max-pooling

	#3
	W_conv_3 = weight_variable([5,5,96,256])#
	b_conv_3 = bias_variable([256])#
	conv2d_3 = conv2d(h_pool_2,W_conv_3,5) + b_conv_3

	#4
	pool_4 = max_pool(conv2d_3,3,2)#进行max-pooling

	#5   第六层是2group，考虑tensorflow切片不方便，这里先输出两组
	W_conv1_5= weight_variable([3,3,256,192])
	b_conv1_5 = bias_variable([192])#
	conv2d1_5 = conv2d(pool_4  ,W_conv1_5,1) + b_conv1_5

	W_conv2_5 = weight_variable([3,3,256,192])
	b_conv2_5 = bias_variable([192])
	conv2d2_5 = conv2d(pool_4 ,W_conv2_5,1) + b_conv2_5

	#6  2group
	W_conv1_6 = weight_variable([3,3,192,192])
	b_conv1_6 = bias_variable([192])
	conv2d1_6 = conv2d(conv2d1_5 ,W_conv1_6,1) + b_conv1_6

	W_conv2_6 = weight_variable([3,3,192,192])
	b_conv2_6 = bias_variable([192])
	conv2d2_6 = conv2d(conv2d2_5 ,W_conv2_6,1) + b_conv2_6

	#7  2group
	W_conv1_7 = weight_variable([3,3,192,128])
	b_conv1_7 = bias_variable([128])
	conv2d1_7 = conv2d(conv2d1_6 ,W_conv1_7,1) + b_conv1_7

	W_conv2_7 = weight_variable([3,3,192,128])
	b_conv2_7 = bias_variable([128])
	conv2d2_7 = conv2d(conv2d2_6 ,W_conv2_7,1) + b_conv2_7

	#8   2group
	pool1_8 = max_pool(conv2d1_7,3,2)#进行max-pooling
	pool2_8 = max_pool(conv2d2_7,3,2)#进行max-pooling
	#拼接
	pool8 = tf.concat([pool1_8, pool2_8],1)   # (10, 4, 2, 128)                        

	W_fc1 = weight_variable([1024,512])#上一场有7*7*64个神经元，全连接层有1024个神经元
	b_fc1 = bias_variable([512])#1024个节点
	h_pool_flat = tf.reshape(pool8,[-1,1024])
	wx_plus_b1 = tf.matmul(h_pool_flat,W_fc1) + b_fc1
	h_fc1 = tf.nn.relu(wx_plus_b1)
	keep_prob = tf.placeholder(tf.float32)
	h_fc1_drop = tf.nn.dropout(h_fc1,keep_prob)

	W_fc2 = weight_variable([512,n_classes])
	b_fc2 = bias_variable([n_classes])
	wx_plus_b2 = tf.matmul(h_fc1_drop,W_fc2) + b_fc2
	prediction = tf.nn.softmax(wx_plus_b2)

	cross_entropy = tf.reduce_mean(tf.nn.softmax_cross_entropy_with_logits(labels=y,logits=prediction),name='cross_entropy')
	correct_prediction = tf.equal(tf.argmax(prediction,1),tf.argmax(y,1))#argmax返回一维张量中最大的值所在的位置
	accuracy = tf.reduce_mean(tf.cast(correct_prediction,tf.float32))  

	train_step = tf.train.AdamOptimizer(1e-4).minimize(cross_entropy)

	with tf.Session() as sess:
		sess.run(tf.global_variables_initializer())
		for epoch in range(num_step):
			#训练模型
			batch_xy = train_p.next_batch(batch_size,shuffle=0)
			batch_xs = removehead(batch_xy)
			batch_ys = one_hot(batch_xy[:,0])
			sess.run(train_step,feed_dict={x:batch_xs,y:batch_ys,keep_prob:0.5})
		  
			if epoch%perprint==0:
				train_acc_allbatch = np.zeros((X_train.shape[0]//batch_size,1))
				for i in range(X_train.shape[0]//batch_size):
					batch_xy = train_p2.next_batch(batch_size,shuffle=0)
					batch_xs = removehead(batch_xy)
					batch_ys = one_hot(batch_xy[:,0])
					train_acc_allbatch[i] = sess.run(accuracy,feed_dict={x:batch_xs,y:batch_ys,keep_prob:0.5})
				train_acc = train_acc_allbatch.mean(axis=0)[0]

				test_acc_allbatch = np.zeros((X_test.shape[0]//batch_size,1))
				for i in range(X_test.shape[0]//batch_size):
					batch_xy = test_p.next_batch(batch_size,shuffle=0)
					batch_xs = removehead(batch_xy)
					batch_ys = one_hot(batch_xy[:,0])
					test_acc_allbatch[i] = sess.run(accuracy,feed_dict={x:batch_xs,y:batch_ys,keep_prob:0.5})
				test_acc = test_acc_allbatch.mean(axis=0)[0]
			
				print ("epoch " + str(epoch) + ", Testing Accuracy= " + str(test_acc) + ", Training Accuracy= " + str(train_acc))
if __name__ == "__main__":
    runModel()
